File: JS/lab4/producer.js
const { Kafka } = require('kafkajs')

const kafka = new Kafka({
  brokers: ['broker-1:9092', 'broker-2:9093'],
  clientId: 'lab4-producer',
})

const producer = kafka.producer()
const typesOfDisasters = ['flood', 'hurricane']

const getIntensity = () => Math.floor(Math.random()* 10)
const createMessage = num => ({
  value: `${num}`,
})

const sendMessage = () => {
  topic = typesOfDisasters[Math.floor(Math.random()*typesOfDisasters.length)];
  return producer
    .send({
      topic,
      messages: Array(1)
        .fill()
        .map(_ => createMessage(getIntensity())),
    })
    .then(console.log)
    .catch(e => console.error(`[example/producer] ${e.message}`, e))
}

const run = async () => {
  await producer.connect()
  setInterval(sendMessage, 3000)
}

run().catch(e => console.error(`[example/producer] ${e.message}`, e))

File: lab4/solutions/src/main/java/com/honolulu/Producer.java
package com.honolulu;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.DoubleSerializer;
import org.apache.kafka.common.serialization.StringSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;
import java.util.Random;

public class Producer {
	private static final Logger log = LoggerFactory.getLogger(Producer.class);

	public static void main(String[] args) {

		Properties props = new Properties();
		props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "http://broker-1:9092,http://broker-2:9093");
		props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
		props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, DoubleSerializer.class.getName());
		KafkaProducer<String, Double> producer = new KafkaProducer<>(props);
		while (true) {
			double value = Math.floor(Math.random()*10+1);
			String[] disasters = {"flood", "hurricane"};
			int idx = new Random().nextInt(disasters.length);
			String key = (disasters[idx]);
			ProducerRecord<String, Double> producerRecord = new ProducerRecord<>(key, key, value);

			log.info("Sending message with key " + key + " to Kafka");

			producer.send(producerRecord, (metadata, e) -> {
				if (metadata != null) {
					System.out.println(producerRecord.key());
					System.out.println(producerRecord.value());
					System.out.println(producerRecord.partition());
					// System.out.println(producerRecord.offset());
				}
			});
			producer.flush();
			log.info("Successfully produced messages to " + key + " topic");
		}
		// producer.close();


	}
}
